File: app/src/main/java/com/example/smartspend/database/TransactionDao.java
// app/src/main/java/com/example/smartspend/database/TransactionDao.java
package com.example.smartspend.database;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import com.example.smartspend.model.Transaction;

import java.util.List;

@Dao
public interface TransactionDao {

    @Insert
    void insert(Transaction transaction);

    @Update
    void update(Transaction transaction);

    @Delete
    void delete(Transaction transaction);

    // Toutes les transactions triées par date décroissante
    @Query("SELECT * FROM transactions WHERE userId = :userId ORDER BY date DESC")
    LiveData<List<Transaction>> getAllTransactions(int userId);

    // Filtre par plage de dates
    @Query("SELECT * FROM transactions WHERE userId = :userId " +
            "AND date BETWEEN :start AND :end ORDER BY date DESC")
    LiveData<List<Transaction>> getTransactionsByDateRange(int userId, long start, long end);

    // Recherche par description ou catégorie
    @Query("SELECT * FROM transactions WHERE userId = :userId " +
            "AND (description LIKE '%' || :query || '%' " +
            "OR category LIKE '%' || :query || '%') ORDER BY date DESC")
    LiveData<List<Transaction>> searchTransactions(int userId, String query);
}

File: app/src/main/java/database/TransactionDao.kt
package com.team.smartspend.database

import androidx.lifecycle.LiveData
import androidx.room.*
import com.team.smartspend.model.Transaction

@Dao
interface TransactionDao {
    @Insert
    suspend fun insert(transaction: Transaction)

    @Update
    suspend fun update(transaction: Transaction)

    @Delete
    suspend fun delete(transaction: Transaction)

    @Query("SELECT * FROM transactions ORDER BY date DESC")
    fun getAllTransactions(): LiveData<List<Transaction>>

    @Query("SELECT * FROM transactions WHERE date BETWEEN :debut AND :fin ORDER BY date DESC")
    fun getTransactionsByDate(debut: Long, fin: Long): LiveData<List<Transaction>>

    @Query("SELECT SUM(montant) FROM transactions WHERE type = 'REVENU'")
    suspend fun getTotalRevenus(): Double?

    @Query("SELECT SUM(montant) FROM transactions WHERE type = 'DEPENSE'")
    suspend fun getTotalDepenses(): Double?

    @Query("SELECT * FROM transactions ORDER BY date DESC LIMIT 5")
    fun getLastTransactions(): LiveData<List<Transaction>>

    // Récupérer une transaction précise (pour l'édition)
    @Query("SELECT * FROM transactions WHERE id = :id LIMIT 1")
    suspend fun getById(id: Int): Transaction?

    // Total des dépenses d'un mois donné (pour la vérification du budget / notifications)
    // :debut et :fin sont les bornes (timestamps) du mois.
    @Query("SELECT SUM(montant) FROM transactions WHERE type = 'DEPENSE' AND date BETWEEN :debut AND :fin")
    suspend fun getTotalDepensesMois(debut: Long, fin: Long): Double?
}